stress: failed test in cockroach/storage/storage.test: TestStoreMetrics

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/c5c09717a13f1436e1d0ceded8c14d2ac86e02d9

Stress build found a failed test:

```
=== RUN   TestStoreMetrics
I160810 04:46:54.593863 storage/engine/rocksdb.go:353  opening in memory rocksdb instance
W160810 04:46:54.594718 gossip/gossip.go:1015  not connected to cluster; use --join to specify a connected node
I160810 04:46:54.596268 storage/store.go:2732  system config not yet available
I160810 04:46:54.596895 storage/replica_trigger.go:308  store=1:1 range=1 [/Min-/Max): new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 900ms following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=1970-01-01 00:00:00 +0000 UTC]
I160810 04:46:54.597123 storage/engine/rocksdb.go:353  opening in memory rocksdb instance
W160810 04:46:54.597713 gossip/gossip.go:1015  not connected to cluster; use --join to specify a connected node
I160810 04:46:54.598183 storage/store.go:2732  system config not yet available
I160810 04:46:54.598225 gossip/client.go:75  starting client to 127.0.0.1:40892
I160810 04:46:54.598393 storage/engine/rocksdb.go:353  opening in memory rocksdb instance
W160810 04:46:54.599262 gossip/gossip.go:1015  not connected to cluster; use --join to specify a connected node
I160810 04:46:54.599861 storage/store.go:2732  system config not yet available
I160810 04:46:54.600008 gossip/client.go:75  starting client to 127.0.0.1:40892
I160810 04:46:54.602263 storage/replica_command.go:2180  store=1:1 range=1 [/Min-/Max): initiating a split of this range at key "m"
E160810 04:46:54.604167 storage/queue.go:508  [replicate] (purgatory) on store=1:1 range=1 [/Min-"m"): 0 of 0 stores with an attribute matching []; likely not enough nodes in cluster
I160810 04:46:54.609357 storage/replica_raftstorage.go:474  generated snapshot for range 2 at index 10 in 179.877µs. encoded size=2407, 23 KV pairs, 0 log entries
I160810 04:46:54.609745 storage/replica_command.go:3023  store=1:1 range=2 ["m"-/Max): change replicas of 2: read existing descriptor range_id:2 start_key:"m" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
I160810 04:46:54.610226 storage/replica_raftstorage.go:600  store=2:2 range=2 [/Min-/Min): with replicaID [?], applying preemptive snapshot for range 2 at index 10 (encoded size=2407, 23 KV pairs, 0 log entries)
I160810 04:46:54.610443 storage/replica_raftstorage.go:603  store=2:2 range=2 ["m"-/Max): with replicaID [?], applied preemptive snapshot for range 2 in 198.621µs
I160810 04:46:54.610631 storage/replica.go:2141  store=1:1 range=1 [/Min-"m"): applied part of replica change txn: Noop [/Min,/Min), Noop [/Min,/Min), Put [/Meta2/Max,/Min), pErr=<nil>
I160810 04:46:54.611297 storage/replica.go:2141  store=1:1 range=2 ["m"-/Max): applied part of replica change txn: BeginTransaction [/Local/Range/"m"/RangeDescriptor,/Min), ConditionalPut [/Local/Range/"m"/RangeDescriptor,/Min), Noop [/Min,/Min), pErr=<nil>
I160810 04:46:54.611542 storage/replica.go:1520  store=1:1 range=2 ["m"-/Max): proposing ADD_REPLICA {NodeID:2 StoreID:2 ReplicaID:2} for range 2: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2}]
I160810 04:46:54.611734 storage/replica.go:2141  store=1:1 range=2 ["m"-/Max): applied part of replica change txn: EndTransaction [/Local/Range/"m"/RangeDescriptor,/Min), pErr=<nil>
I160810 04:46:54.612545 storage/replica_raftstorage.go:474  generated snapshot for range 2 at index 13 in 192.332µs. encoded size=3231, 25 KV pairs, 3 log entries
W160810 04:46:54.612842 storage/intent_resolver.go:344  could not GC completed transaction: end key /Local/Range/"m"/"rdsc\x00" must be greater than start /Local/Range/"m"/RangeDescriptor
I160810 04:46:54.613037 storage/replica_command.go:3023  store=1:1 range=2 ["m"-/Max): change replicas of 2: read existing descriptor range_id:2 start_key:"m" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > replicas:<node_id:2 store_id:2 replica_id:2 > next_replica_id:3 
I160810 04:46:54.613487 storage/replica.go:2141  store=1:1 range=1 [/Min-"m"): applied part of replica change txn: Noop [/Min,/Min), Noop [/Min,/Min), Put [/Meta2/Max,/Min), pErr=<nil>
I160810 04:46:54.614002 storage/replica_raftstorage.go:600  store=3:3 range=2 [/Min-/Min): with replicaID [?], applying preemptive snapshot for range 2 at index 13 (encoded size=3231, 25 KV pairs, 3 log entries)
I160810 04:46:54.614773 storage/replica_raftstorage.go:603  store=3:3 range=2 ["m"-/Max): with replicaID [?], applied preemptive snapshot for range 2 in 718.655µs
I160810 04:46:54.615523 storage/replica.go:2141  store=2:2 range=2 ["m"-/Max): applied part of replica change txn: BeginTransaction [/Local/Range/"m"/RangeDescriptor,/Min), ConditionalPut [/Local/Range/"m"/RangeDescriptor,/Min), Noop [/Min,/Min), pErr=<nil>
I160810 04:46:54.615729 storage/replica.go:2141  store=1:1 range=2 ["m"-/Max): applied part of replica change txn: BeginTransaction [/Local/Range/"m"/RangeDescriptor,/Min), ConditionalPut [/Local/Range/"m"/RangeDescriptor,/Min), Noop [/Min,/Min), pErr=<nil>
I160810 04:46:54.615749 storage/replica.go:2141  store=2:2 range=2 ["m"-/Max): applied part of replica change txn: EndTransaction [/Local/Range/"m"/RangeDescriptor,/Min), pErr=<nil>
I160810 04:46:54.616099 storage/replica.go:1520  store=1:1 range=2 ["m"-/Max): proposing ADD_REPLICA {NodeID:3 StoreID:3 ReplicaID:3} for range 2: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:2 StoreID:2 ReplicaID:2} {NodeID:3 StoreID:3 ReplicaID:3}]
I160810 04:46:54.616255 storage/replica.go:2141  store=2:2 range=2 ["m"-/Max): applied part of replica change txn: BeginTransaction [/Local/Range/"m"/RangeDescriptor,/Min), ConditionalPut [/Local/Range/"m"/RangeDescriptor,/Min), Noop [/Min,/Min), pErr=<nil>
I160810 04:46:54.616930 storage/replica.go:2141  store=1:1 range=2 ["m"-/Max): applied part of replica change txn: EndTransaction [/Local/Range/"m"/RangeDescriptor,/Min), pErr=<nil>
I160810 04:46:54.617186 stopper.go:408  quiesceing; tasks left:
1      storage/intent_resolver.go:346
1      kv/txn_coord_sender.go:873
I160810 04:46:54.617265 storage/replica.go:2141  store=2:2 range=2 ["m"-/Max): applied part of replica change txn: EndTransaction [/Local/Range/"m"/RangeDescriptor,/Min), pErr=<nil>
W160810 04:46:54.617277 storage/intent_resolver.go:325  store=1:1 range=2 ["m"-/Max): failed to resolve intents: failed to send RPC: sending to all 1 replicas failed; last error: failed to send RPC: store is stopped
I160810 04:46:54.617301 stopper.go:408  quiesceing; tasks left:
1      kv/txn_coord_sender.go:873
W160810 04:46:54.617485 storage/replica.go:410  store=2:2 range=2 ["m"-/Max): outgoing raft transport stream to {1 1 1} closed by the remote: rpc error: code = 2 desc = unable to accept Raft message from {NodeID:1 StoreID:1 ReplicaID:1}: no store registered for {NodeID:2 StoreID:2 ReplicaID:2}
I160810 04:46:55.021304 storage/replica_raftstorage.go:474  generated snapshot for range 2 at index 15 in 200.032µs. encoded size=4094, 27 KV pairs, 5 log entries
I160810 04:46:55.022065 storage/replica_raftstorage.go:600  store=3:3 range=2 [/Min-/Min): with replicaID 3, applying Raft snapshot for range 2 at index 15 (encoded size=4094, 27 KV pairs, 5 log entries)
I160810 04:46:55.022409 storage/replica_raftstorage.go:603  store=3:3 range=2 ["m"-/Max): with replicaID 3, applied Raft snapshot for range 2 in 316.862µs
I160810 04:46:55.022497 storage/replica_trigger.go:308  store=2:2 range=2 ["m"-/Max): new range lease replica {2 2 2} 1970-01-01 00:00:00.9 +0000 UTC 1.8s following replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 900ms [physicalTime=1970-01-01 00:00:01.8 +0000 UTC]
I160810 04:46:55.025177 stopper.go:408  quiesceing; tasks left:
1      storage/intent_resolver.go:306
W160810 04:46:55.025330 storage/intent_resolver.go:299  store=1:1 range=1 [/Min-"m"): failed to resolve intents: node unavailable; try another peer
W160810 04:46:55.025728 storage/replica.go:410  store=1:1 range=2 ["m"-/Max): outgoing raft transport stream to {2 2 2} closed by the remote: rpc error: code = 2 desc = unable to accept Raft message from {NodeID:2 StoreID:2 ReplicaID:2}: no store registered for {NodeID:1 StoreID:1 ReplicaID:1}
I160810 04:46:55.025756 http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:42392->127.0.0.1:52878: use of closed network connection
I160810 04:46:55.025775 http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:40892->127.0.0.1:36576: use of closed network connection
I160810 04:46:55.025787 http2_server.go:276  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:40610->127.0.0.1:45320: use of closed network connection
--- FAIL: TestStoreMetrics (0.44s)
	client_metrics_test.go:92: cluster_id:<00000000-0000-0000-0000-000000000000> node_id:1 store_id:1 : expected intent count to be zero, was 1
```

Run Details:

```
0 runs so far, 0 failures, over 5s
0 runs so far, 0 failures, over 10s
0 runs so far, 0 failures, over 15s
0 runs so far, 0 failures, over 20s
0 runs so far, 0 failures, over 25s
0 runs so far, 0 failures, over 30s
1 runs so far, 0 failures, over 35s
2 runs completed, 1 failures, over 37s
FAIL
```
Please assign, take a look and update the issue accordingly.

Git Bisect shows two suspicious commits

d41dc7240213bfe764336b54139ec3bb872daf99 is the first commit that yields this particular failure (intent count 1, should be 0).

bee9130bbbcc56c2bc9d2a789708f6bb7fe34404 actually produces an unrelated panic:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x110 pc=0x411ae15]

goroutine 715 [running]:
panic(0x5008a80, 0xc420012100)
	/Users/matt/golang17/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/util/stop.(*Stopper).Recover(0xc4205c5200)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:176 +0x6e
panic(0x5008a80, 0xc420012100)
	/Users/matt/golang17/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/storage.LeaseExpiration(0x0, 0xc420152080, 0xc42116e601)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/storage/replica_test.go:107 +0x35
github.com/cockroachdb/cockroach/storage_test.(*multiTestContext).expireLeases(0xc420001860)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/storage/client_test.go:862 +0x4a
github.com/cockroachdb/cockroach/storage_test.(*multiTestContextKVTransport).SendNext(0xc4200b3a20, 0xc4209d0420)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/storage/client_test.go:412 +0x19f
github.com/cockroachdb/cockroach/kv.(*DistSender).sendToReplicas(0xc4205c5290, 0x7c56bb8, 0xc421180000, 0x2540be400, 0xb2d05e00, 0xc42014bf40, 0x2, 0xc420978570, 0x2, 0x2, ...)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/kv/dist_sender.go:977 +0x71e
github.com/cockroachdb/cockroach/kv.(*DistSender).sendRPC(0xc4205c5290, 0x7c56bb8, 0xc421180000, 0x2, 0xc420978570, 0x2, 0x2, 0x0, 0x0, 0x0, ...)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/kv/dist_sender.go:364 +0x1c6
github.com/cockroachdb/cockroach/kv.(*DistSender).sendSingleRange(0xc4205c5290, 0x7c56bb8, 0xc421180000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/kv/dist_sender.go:480 +0x19f
github.com/cockroachdb/cockroach/kv.(*DistSender).sendChunk.func2(0xc4204669c8, 0xc420466868, 0xc4204670b0, 0xc4205c5290, 0xc4204670a0, 0xd, 0xc4204bfeb3)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/kv/dist_sender.go:739 +0x18e
github.com/cockroachdb/cockroach/kv.(*DistSender).sendChunk(0xc4205c5290, 0x7c56bb8, 0xc421180000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/kv/dist_sender.go:740 +0x437
github.com/cockroachdb/cockroach/kv.(*DistSender).Send(0xc4205c5290, 0x7c56bb8, 0xc421180000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/kv/dist_sender.go:584 +0x3c0
github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).Send(0xc420074b40, 0x7c56bb8, 0xc421180000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:375 +0x546
github.com/cockroachdb/cockroach/internal/client.(*DB).send(0xc4202373b0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/internal/client/db.go:486 +0x220
github.com/cockroachdb/cockroach/internal/client.(*DB).(github.com/cockroachdb/cockroach/internal/client.send)-fm(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/internal/client/db.go:433 +0x58
github.com/cockroachdb/cockroach/internal/client.sendAndFill(0xc420467ce0, 0xc42117e000, 0x0, 0xc420467e20)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/internal/client/db.go:415 +0x16b
github.com/cockroachdb/cockroach/internal/client.(*DB).Run(0xc4202373b0, 0xc42117e000, 0x1, 0x1)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/internal/client/db.go:433 +0x75
github.com/cockroachdb/cockroach/storage.(*intentResolver).processIntentsAsync.func2()
	/Users/matt/go/src/github.com/cockroachdb/cockroach/storage/intent_resolver.go:346 +0x586
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunLimitedAsyncTask.func1(0xc4205c5200, 0x576f6ad, 0x1a, 0x15d, 0xc4205996e0, 0xc420381c80)
	/Users/matt/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:269 +0xbf
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunLimitedAsyncTask
	/Users/matt/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:270 +0x1a7
```

Investigating both to figure out the cause of this issue.

--- a/storage/client_metrics_test.go
+++ b/storage/client_metrics_test.go
@@ -17,6 +17,7 @@
 package storage_test
 
 import (
+	"fmt"
 	"sync"
 	"testing"
 
@@ -25,6 +26,7 @@ import (
 	"github.com/cockroachdb/cockroach/internal/client"
 	"github.com/cockroachdb/cockroach/roachpb"
 	"github.com/cockroachdb/cockroach/storage"
+	"github.com/cockroachdb/cockroach/util"
 	"github.com/cockroachdb/cockroach/util/leaktest"
 	"github.com/cockroachdb/cockroach/util/metric"
 	"github.com/pkg/errors"
@@ -59,6 +61,19 @@ func verifyStats(t *testing.T, mtc *multiTestContext, storeIdxSlice ...int) {
 	}
 	mtc.mu.RUnlock()
 
+	// Sanity regression check for bug #4624: ensure intent count is zero.
+	// This may not be true immediately due to the asynchronous nature of
+	// non-local intent resolution.
+	for _, s := range stores {
+		m := s.Metrics()
+		util.SucceedsSoon(t, func() error {
+			if a := m.IntentCount.Value(); a != 0 {
+				return fmt.Errorf("expected intent count to be zero, was %d", a)
+			}
+			return nil
+		})
+	}
+
 	wg.Add(numStores)
 	// We actually stop *all* of the Stores. Stopping only a few is riddled
 	// with deadlocks since operations can span nodes, but stoppers don't
@@ -78,22 +93,19 @@ func verifyStats(t *testing.T, mtc *multiTestContext, storeIdxSlice ...int) {
 			prefix := s.Ident.String() + ": "
 			t.Fatalf(prefix+msg, args...)
 		}
-		// Compute real total MVCC statistics from store.
-		realStats, err := s.ComputeMVCCStats()
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		// Sanity regression check for bug #4624: ensure intent count is zero.
-		if a := realStats.IntentCount; a != 0 {
-			fatalf("expected intent count to be zero, was %d", a)
-		}
 
 		m := s.Metrics()
+
 		// Sanity check: LiveBytes is not zero (ensures we don't have
 		// zeroed out structures.)
 		if liveBytes := m.LiveBytes.Value(); liveBytes == 0 {
 			fatalf("expected livebytes to be non-zero, was zero")
+		}
+
+		// Compute real total MVCC statistics from store.
+		realStats, err := s.ComputeMVCCStats()
+		if err != nil {
+			t.Fatal(err)
 		}
 
 		// Ensure that real MVCC stats match computed stats.
@@ -154,7 +166,8 @@ func verifyRocksDBStats(t *testing.T, s *storage.Store) {
 
 func TestStoreMetrics(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	t.Skip("TODO(mrtracy): #8437")
+	t.Skip("TODO(mrtracy): #9204")
+
 	mtc := startMultiTestContext(t, 3)
 	defer mtc.Stop()
 

